### PR TITLE
Cheat sheet for initial review "Microservices-based Security Architecture Documentation"

### PR DESCRIPTION
**Pull request by AlexB1986**
Cheat sheet for the initial review "Microservices-based Security Architecture Documentation":
- [x] In case of a new Cheat Sheet, you have used the [Cheat Sheet template](https://github.com/OWASP/CheatSheetSeries/blob/master/templates/New_CheatSheet.md).
- [x] All the markdown files do not raise any validation policy violation, see policy [here](https://github.com/OWASP/CheatSheetSeries#editor--validation-policy).
- [x] All the markdown files follow these [format rules](https://github.com/OWASP/CheatSheetSeries#conversion-rules).
- [x] All your assets are stored in the **assets** folder.
- [x] All the images used are in the **PNG** format.
- [x] Any references to website have been formatted as [TEXT](URL)
- [x] You verified/tested the effectiveness of your contribution (e.g.: defensive code proposed is really an effective remediation? Please verify it works!).
- [x] The CI build of your PR pass, see the build status [here](https://travis-ci.org/OWASP/CheatSheetSeries/pull_requests).

This PR covers issue #40 .

### CODE REVIEW

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+## Create a graphical presentation of application architecture
+
+It is advisable to create graphical presentation of application architecture (building blocks and relations defined above) in form of services call graph or data flow diagram. In order to do that one can use special software tools (e.g. Enterprise Architect) or [DOT language](https://en.wikipedia.org/wiki/DOT_(graph_description_language)). See example of using DOT language [here](https://gist.github.com/vladgolubev/80c5523336ddec3859c0e90d9a070882).
+
+## Use collected information in secure software development practices
+
+Collected information may be useful for doing application security practices, e.g. during defining security requirements, threat modeling or security testing. Table below contains examples of activities related to securing application architecture (as well as its mapping to OWASP ASVS 4.0) and tips for their implementation using information collected above.
+
+| Activity description | Implementation tips | Mapping to OWASP ASVS 4.0 |
```
Comment (mackowski) [inline cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: I am wondering if the table is the best option here. For me 'Implementation tips' are poorly readable. 
One way to improve readability is to use bullet points inside table e.g.

| Activity description | Implementation tips | Mapping to OWASP ASVS 4.0 |
| :--- | :--- | :--- |
| Attack surface analysis | To enumerate microservices endpoints that need to be tested during security testing and analyzed during threat modeling analyze data collected under the following sections: <ul><li>"Identify and describe application-functionality services" (parameter “API definition”) </li><li>"Identify and describe infrastructure services" (parameter “Link to the service documentation”)</li></ul> | 1.1.2

But I do not know how to do this without HTML syntax inside table...

Another approach will be to refactor it to other form than table.
What do you think @righettod @AlexB1986 @ThunderSon

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
+# Introduction
+
+The microservice architecture is being increasingly used for designing and implementing application systems in both cloud-based and enterprise infrastructures. There are many security challenges need to be addressed in the application design and implementation phases. In order to address some security challenges it is necessity to collect security-specific information on application architecture.
+The goal of this article is to provide a concrete proposal of approach to collect microservice-based architecture information to securing application.
+
+# Context
+
+During securing applications based on microservices architecture, security architects/engineers usually face with the following questions (mostly referenced in the OWASP Application Security Verification Standard Project under the section V1 “Architecture, Design and Threat Modeling Requirements”):
```
Comment (mackowski) [inline cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Add links to ASVS e.g.

mostly referenced in the [OWASP Application Security Verification Standard Project](https://github.com/OWASP/ASVS) under the section
[V1 “Architecture, Design and Threat Modeling Requirements”](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x10-V1-Architecture.md#v1-architecture-design-and-threat-modeling-requirements)

**Comment by mackowski** [comment]
Thank you for this PR :)

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
+# Introduction
+
+The microservice architecture is being increasingly used for designing and implementing application systems in both cloud-based and enterprise infrastructures. There are many security challenges need to be addressed in the application design and implementation phases. In order to address some security challenges it is necessity to collect security-specific information on application architecture.
```
[the remark this answers is not in the document]
Comment (AlexB1986) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Yes, fixed.

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md (hunk 2 above)
Comment (AlexB1986) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Fixed

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md (hunk 1 above)
Comment (AlexB1986) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Thanks for your time and review!

Yes, my first version (Word) was with that bullet points.
We may refactor it in the following way:


## Use collected information in secure software development practices

....
### Attack surface analysis
#### Implementation tips
To enumerate microservices endpoints that need to be tested during security testing and analyzed during threat modeling analyze data collected under the following sections:
* "Identify and describe application-functionality services" (parameter “API definition”);
* "Identify and describe infrastructure services" (parameter “Link to the service documentation”).
#### Mapping to OWASP projects
* [OWASP ASVS, V1 “Architecture, Design and Threat Modeling Requirements”, #1.1.2](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x10-V1-Architecture.md#v1-architecture-design-and-threat-modeling-requirements)
* [OWASP Attack Surface Analysis Cheat Sheet](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Attack_Surface_Analysis_Cheat_Sheet.md)

Comment (righettod) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: I think like you guys, table is not well suitable for this kind of formatting of content.
The refactoring proposed by @AlexB1986 is better.
We only use HTML in very rare condition when there is no other way to achieve an expected result.

Comment (mackowski) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: @AlexB1986 I also like the formatting that you proposed :)

Comment (AlexB1986) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: OK, will update it soon

Comment (AlexB1986) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Done

**Comment by righettod** [comment]
I have added [info](https://github.com/OWASP/CheatSheetSeries/issues/105#issuecomment-492650331) about the TravisCI job issue to fix it in existing PR.

**Comment by AlexB1986** [comment]
@righettod Are there any issues with that PR?
Thanks

**Comment by righettod** [comment]
Hello,

We wait the review from @mackowski and @thunderson .

Regards,
Dominique

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+To verify the definition and documentation of all application components in terms of the business or security functions they provide analyze data collected under the following sections (parameter “Short description”): 
+* "Identify and describe application-functionality services";
+* "Identify and describe infrastructure services".
+
+#### Mapping to OWASP projects
+
+* [OWASP ASVS, V1 “Architecture, Design and Threat Modeling Requirements”, #1.11.1](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x10-V1-Architecture.md#v1-architecture-design-and-threat-modeling-requirements)
+
+# Authors and Primary Editors
```
Comment (ThunderSon) [inline cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Kindly remove this section as we are relying on the Github functionality for the contributors

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+* [OWASP ASVS, V1 “Architecture, Design and Threat Modeling Requirements”, #1.1.2](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x10-V1-Architecture.md#v1-architecture-design-and-threat-modeling-requirements)
+* [OWASP Attack Surface Analysis Cheat Sheet](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Attack_Surface_Analysis_Cheat_Sheet.md)
+
+### Data leakage analysis
+
+#### Implementation tips
+
+To analyze possible data leakage analyze data collected under the following sections: 
+* "Identify and describe data assets";
```
Comment (ThunderSon) [inline cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Clean the bulleted points. Remove the `;` and the `" "` surrounding them

Comment (AlexB1986) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Done

File: cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md (hunk 4 above)
Comment (AlexB1986) [reply cheatsheets/Microservices_based_Security_Arch_Doc_Cheat_Sheet.md]: Done

**Pull request merged**